File: benchmark/comparison/results/blind_py/22.py
class Lit:
    def __init__(self, n):
        self.n = n

class Add:
    def __init__(self, a, b):
        self.a = a
        self.b = b

class Mul:
    def __init__(self, a, b):
        self.a = a
        self.b = b

def eval_expr(e):
    if isinstance(e, Lit):
        return e.n
    elif isinstance(e, Add):
        return eval_expr(e.a) + eval_expr(e.b)
    elif isinstance(e, Mul):
        return eval_expr(e.a) * eval_expr(e.b)

def describe(e):
    if isinstance(e, Lit):
        return str(e.n)
    elif isinstance(e, Add):
        return f"{describe(e.a)} + {describe(e.b)}"
    elif isinstance(e, Mul):
        return f"{describe(e.a)} * {describe(e.b)}"

for e in [Lit(5), Add(Lit(3), Lit(4)), Mul(Lit(3), Lit(4))]:
    print(eval_expr(e))
    print(describe(e))

File: benchmark/comparison/results/blind_ts/22.ts
type Expr =
  | { type: "Lit"; value: number }
  | { type: "Add"; a: Expr; b: Expr }
  | { type: "Mul"; a: Expr; b: Expr };

function Lit(n: number): Expr { return { type: "Lit", value: n }; }
function Add(a: Expr, b: Expr): Expr { return { type: "Add", a, b }; }
function Mul(a: Expr, b: Expr): Expr { return { type: "Mul", a, b }; }

function eval_expr(e: Expr): number {
  switch (e.type) {
    case "Lit": return e.value;
    case "Add": return eval_expr(e.a) + eval_expr(e.b);
    case "Mul": return eval_expr(e.a) * eval_expr(e.b);
  }
}

function describe(e: Expr): string {
  switch (e.type) {
    case "Lit": return String(e.value);
    case "Add": return `${describe(e.a)} + ${describe(e.b)}`;
    case "Mul": return `${describe(e.a)} * ${describe(e.b)}`;
  }
}

console.log(eval_expr(Lit(5)));
console.log(eval_expr(Add(Lit(3), Lit(4))));
console.log(eval_expr(Mul(Lit(3), Lit(4))));
console.log(describe(Lit(5)));
console.log(describe(Add(Lit(3), Lit(4))));
console.log(describe(Mul(Lit(3), Lit(4))));
